src/Type/Php/CallUserFuncDynamicReturnTypeExtension.php

```diff
@@ -0,0 +1,33 @@
+<?php declare(strict_types = 1);
+
+namespace PHPStan\Type\Php;
+
+use PhpParser\Node\Expr\FuncCall;
+use PHPStan\Analyser\Scope;
+use PHPStan\Reflection\FunctionReflection;
+use PHPStan\Reflection\ParametersAcceptorSelector;
+use PHPStan\Type\DynamicFunctionReturnTypeExtension;
+use PHPStan\Type\Type;
+
+class CallUserFuncDynamicReturnTypeExtension implements DynamicFunctionReturnTypeExtension
+{
+
+	public function isFunctionSupported(FunctionReflection $functionReflection): bool
+	{
+		return $functionReflection->getName() === 'call_user_func';
+	}
+
+	public function getTypeFromFunctionCall(FunctionReflection $functionReflection, FuncCall $functionCall, Scope $scope): Type
+	{
+		$args = $functionCall->getArgs();
+		$defaultReturn = ParametersAcceptorSelector::selectSingle($functionReflection->getVariants())->getReturnType();
+
+		$callbackType = $scope->getType($args[0]->value);
+		if ($callbackType->isCallable()->yes()) {
+			return $scope->getType(new FuncCall($args[0]->value, array_slice($args, 1)));
+		}
```

Comment:
With arguments provided, would this resolve the second half of my concerns in https://github.com/phpstan/phpstan/issues/5934?

With the arguments provided to `FuncCall`, will we get validation of the parameters passed to the function against its actual signature?

Reply:
Named parameters also require https://github.com/phpstan/phpstan-src/pull/750 as a first step

IIRC I tried adding parameter types signature testing, but removed it because it wasn‘t able to make it work (figured it could be a followup, as the PR as is already fixed a open issue)

Reply:
Makes sense :) No need to make this heavier than it already is, sounds like a great follow up.

Reply:
Checking parameter types for these functions requires a new rule that will utilize FunctionCallParametersCheck.

Reply:
I would love we could land the PR as is and add a new rule after it...?